validator/src/bootstrap.rs

```diff
@@ -790,19 +790,66 @@ fn get_snapshot_hashes_from_known_validators(
+        known_validators,
+        get_full_snapshot_hashes_for_node,
+        get_incremental_snapshot_hashes_for_node,
+        incremental_snapshot_fetch,
+    ) {
+        debug!(
+            "Not all snapshot hashes have been discovered from known validators. \
+            This likely means the gossip tables are not fully populated. \
+            We will sleep and retry..."
+        );
+        return KnownSnapshotHashes::default();
+    }
+
+    build_known_snapshot_hashes(
+        known_validators,
+        get_full_snapshot_hashes_for_node,
+        get_incremental_snapshot_hashes_for_node,
+        incremental_snapshot_fetch,
+    )
+}
+
+/// Check if we can discover all snapshot hashes for all the known validators.
+///
+/// This is a work-around to ensure the gossip tables are populated enough so that the bootstrap
+/// process will download both full and incremental snapshots.  If the incremental snapshot hashes
+/// are not yet populated from gossip, then it is possible (and has been seen often) to only
+/// discover full snapshots—and ones that are very old (up to 25,000 slots)—but *not* discover any
+/// of their associated incremental snapshots.
+///
+/// This function will return false if we do not yet have snapshot hashes from any known validator;
+/// and true otherwise.
+fn do_known_validators_have_all_snapshot_hashes<'a, F1, F2>(
+    known_validators: impl IntoIterator<Item = &'a Pubkey>,
+    get_full_snapshot_hashes_for_node: F1,
+    get_incremental_snapshot_hashes_for_node: F2,
+    incremental_snapshot_fetch: bool,
+) -> bool
+where
+    F1: Fn(&'a Pubkey) -> Vec<(Slot, Hash)>,
+    F2: Fn(&'a Pubkey) -> Option<((Slot, Hash), Vec<(Slot, Hash)>)>,
+{
+    let node_has_full_snapshot_hashes = |node| !get_full_snapshot_hashes_for_node(node).is_empty();
+    let node_has_incremental_snapshot_hashes = |node| {
+        get_incremental_snapshot_hashes_for_node(node)
+            .map(|(_, hashes)| !hashes.is_empty())
+            .unwrap_or(false)
+    };
+
+    // Does this node have all the snapshot hashes?
+    // If incremental snapshots are disabled, only check for full snapshot hashes; otherwise check
+    // for both full and incremental snapshot hashes.
+    let node_has_all_snapshot_hashes = |node| {
+        node_has_full_snapshot_hashes(node)
+            && (!incremental_snapshot_fetch || node_has_incremental_snapshot_hashes(node))
+    };
+
     known_validators
-        .map(|known_validators| {
-            build_known_snapshot_hashes(
-                known_validators,
-                get_full_snapshot_hashes_for_node,
-                get_incremental_snapshot_hashes_for_node,
-                incremental_snapshot_fetch,
-            )
-        })
-        .unwrap_or_else(|| {
-            trace!("No known validators, so no known snapshot hashes");
-            KnownSnapshotHashes::new()
-        })
+        .into_iter()
+        .all(node_has_all_snapshot_hashes)
```

Comment:
m.b. `any`? Otherwise if one of known validators is down or don't produce incremental snapshots then the node will never start.

But the same result will be received just after new full snapshot has been created (normal cluster work). So and `all` and `any` lays a potentially vulnerability in the event of a massive problem - the cluster will not be able to restart automatically within the some time.

It seems to me that to wait for the gossip, you can simply add a default 10 second delay, which will be managed by a new argument. If this closes 90% of the cases, then for a very simple change it looks very good.

In general, I'm leaning towards the solution I [suggested](https://github.com/solana-labs/solana/pull/28143#issuecomment-1262941899): don't concentrate on complex logic in the bootstrap process (it makes the code much more complicated), just check after bootstrap, and restart bootstrap if needed.
I'm sorry, I have no inspiration to suggest this in code myself. But I remember, I'll try to find it.

Reply:
In regards to the `any()` v. `all()` approach that @diman-io called out, @brooksprumo and I were chatting and thought some sort of hybrid change might be interesting to consider. Perhaps try for `all()` but then settle for `any()` after a failed attempt or two to get `all()`.

This allows things to work well in the good case (all known validators up), but avoids getting stuck should a known validator be down. We would definitely hope that validators are attentive as to who their known validators are, but writing the code to avoid a potential foot-gun seems prudent as well. Especially if someone would have to modify their startup script to temporarily remove the known validator to get past this